Add chunksize param to read_json when lines=True

louispotok opened the pull request:
Previous behavior: reading the whole file to memory and then split into lines.
New behavior: if `lines=True` and `chunksize` is passed: read in `chunksize` lines at a time, and concat.

This only covers some kinds of input to `read_json`.  When `chunksize` is passed, `read_json` becomes slower but more memory-efficient.

Closes #17048.

Tests and style-check pass, no new tests added.

louispotok committed:
Add chunksize param to read_json when lines=True

Previous behavior: reading the whole file to memory and then split into lines.
New behavior, if lines=True and chunksize is passed: read in `chunksize`
lines at a time, and concat.
This only covers some kinds of input to read_json.
When chunksize is passed, read_json becomes slower but more
memory-efficient.

gfyoung commented:
@louispotok : Thanks for doing this!  Before we can merge, you will need to add tests and `whatsnew` entry to this PR.

gfyoung commented on pandas/io/json/json.py:
```diff
@@ -263,6 +265,14 @@ def read_json(path_or_buf=None, orient=None, typ='frame', dtype=True,
+        Also note this is different from the `chunksize` parameter in
+            `read_csv`, which returns a FileTextReader.
+        If the JSON input is a string, this argument has no effect.
+
```
Add a version-added tag.

louispotok replied:
Will do.

gfyoung commented on pandas/io/json/json.py:
```diff
@@ -322,6 +332,27 @@ def read_json(path_or_buf=None, orient=None, typ='frame', dtype=True,
     filepath_or_buffer, _, _ = get_filepath_or_buffer(path_or_buf,
                                                       encoding=encoding)
+
+    def _read_json_as_lines(fh, chunksize):
```
Add a docstring for this.  Will be useful for developers.

louispotok replied:
Will do.

gfyoung commented on pandas/io/json/json.py:
```diff
@@ -349,6 +386,13 @@ def read_json(path_or_buf=None, orient=None, typ='frame', dtype=True,
+
+def _get_obj(typ, json, orient, dtype, convert_axes, convert_dates,
+             keep_default_dates, numpy, precise_float,
+             date_unit):
```
Why did you do this?  I like abstraction but not sure how this relates to your PR.

louispotok replied:
Before, this code was only called once. Now it needs to be called from two places, so I extracted it into a function (see line 343). 

One possible improvement I was considering is to define the function inline in the read_json function so that we don't need to pass so many parameters.

jreback commented:
needs tests that validate chunksize is current and errors correctly

see how read_csv does this

louispotok committed:
Add read_json chunksize change to whatsnew

louispotok committed:
Add versionadded to docstring

louispotok committed:
add docstring for _read_json_as_lines

louispotok committed:
add basic read_json chunksize test

louispotok committed:
validate read_json chunksize is an integer and >=1

louispotok commented:
@jreback thanks for the tip. I looked at the read_csv tests and I'm still a bit confused. Specifically I'm looking at "test_read_chunksize" in tests/io/parser/common.py. I see that it's testing two main things:
1) Each returned chunk is equal to that slice of the input dataframe -- not relevant in this case, since read_json is still returning the whole dataframe.
2) Ensure that read_csv is using _validate_integer. I wasn't using that on read_json, but I added it  and am now testing for it.

louispotok committed:
Add more tests to read_json chunksize

- Errors correctly when non-integer is passed as chunksize
- Accepts a float that is close to an integer as chunksize

louispotok commented:
Okay, the latest commits should address all these comments:

- describe change in whatsnew
- add version-added tag to the chunksize param in the read_json docstring
- add docstring for _read_json_as_lines internal function
- validate that chunksize is integer-like and >=1
- add tests for chunksize: reads correctly and rejects invalid chunksize argument.

louispotok commented:
Also, do you prefer to leave these atomic commits or should I squash them?

gfyoung commented:
@louispotok : Whatever is most convenient for you.  We'll squash them at the end once we merge.

louispotok commented:
Thanks @gfyoung , this is ready for another round of review.

gfyoung commented:
@louispotok : This looks pretty good (there are some minor stylistic things I think...@jreback?)

Could you post timing information so that we can see the performance difference with and without your changes in this PR?

jorisvandenbossche commented on doc/source/whatsnew/v0.21.0.txt:
```diff
@@ -197,6 +197,7 @@ Other API Changes
   raise an ``AttributeError`` rather than a ``ClosedFileError`` (:issue:`16301`)
 - :func:`read_csv` now treats ``'null'`` strings as missing values by default (:issue:`16471`)
 - :func:`read_csv` now treats ``'n/a'`` strings as missing values by default (:issue:`16078`)
+- :func:`read_json` now accepts a ``chunksize`` parameter that can reduce memory usage when ``lines=True``. (:issue:`17048`)
```
Can you put this in the "Other Enhancements" section?

jreback replied:
update for this

jorisvandenbossche commented on pandas/io/json/json.py:
```diff
@@ -263,6 +266,16 @@ def read_json(path_or_buf=None, orient=None, typ='frame', dtype=True,
+        If this is None, the file will be read into memory all at once.
+        Passing a chunksize helps with memory usage, but is slower.
+        Also note this is different from the `chunksize` parameter in
+            `read_csv`, which returns a FileTextReader.
```
Shouldn't our goal be to have it with a similar behaviour?

(formatting comment: there is some extra indentation on this line that is not needed

jreback replied:
need to update these doc-strings

jorisvandenbossche commented on pandas/io/json/json.py:
```diff
@@ -263,6 +266,16 @@ def read_json(path_or_buf=None, orient=None, typ='frame', dtype=True,
         .. versionadded:: 0.19.0
 
+    chunksize: integer, default None
+        If `lines=True`, how many lines to read into memory at a time.
```
best to use double backticks here (so ``` ``lines=True`` ```), then it will be formatted as code

louispotok replied:
Will do.

jorisvandenbossche reviewed (commented):
Should we raise an error when a value for `chunksize` is passed and `lines` is not True ?

jorisvandenbossche commented:
My main comment/question here is: should the API be consistent with how `chunksize` is for `read_csv` ? 
This would mean that when using this, the return value is an iterator.

@louispotok I understand that you don't need it in your specific case (as once concatenated, your dataframe fits into memory). But in general it could be needed in certain cases to have it as an iterator. For your case you can then do the concat yourself:

```
In [41]: s = """a,b
    ...: 1,2
    ...: 3,4
    ...: 5,6
    ...: 7,8"""

In [42]: pd.read_csv(StringIO(s))
Out[42]: 
   a  b
0  1  2
1  3  4
2  5  6
3  7  8

In [43]: pd.read_csv(StringIO(s), chunksize=2)
Out[43]: <pandas.io.parsers.TextFileReader at 0x7f2e9c2ef630>

In [44]: pd.concat(pd.read_csv(StringIO(s), chunksize=2))
Out[44]: 
   a  b
0  1  2
1  3  4
2  5  6
3  7  8
```

jreback commented on pandas/io/json/json.py:
```diff
@@ -263,6 +266,16 @@ def read_json(path_or_buf=None, orient=None, typ='frame', dtype=True,
+        If this is None, the file will be read into memory all at once.
+        Passing a chunksize helps with memory usage, but is slower.
+        Also note this is different from the `chunksize` parameter in
+            `read_csv`, which returns a FileTextReader.
```
you need to return an iterator (as @jorisvandenbossche indicates).

you return a new class that inherits from ``BaseIterator`` (in pandas.io.common); see how ``pandas.io.stata.StateReader`` does this). You pretty much just return a class and define ``__next__``, which processes the chunked rows (also an example of this in pandas.io.pytables, though that doesn't use ``BaseIterator``, it should, which is another issue)

jorisvandenbossche replied:
Is this needed? (the subclassing from BaseIterator) We could also just do a `yield` of the chunk in the `_read_json_as_lines` function of the PR (instead of returning the concatted frame of all chunks) 
That seems less invasive to add given the current implementation of `read_json`

jreback replied:
yes because that's how its done elsewhere, and it avoid weird things like py 2/3 compat issues (this is in fact the point of BaseIterator).

jorisvandenbossche replied:
What is the python 2/3 compat issue with using `yield` ? (it has no `.next` method in python 3, but you should use `next(..)` anyway?) 
(that's how it is done in `read_sql`, so if there is an issue it should be fixed there)

jreback commented on pandas/io/json/json.py:
```diff
@@ -322,6 +335,39 @@ def read_json(path_or_buf=None, orient=None, typ='frame', dtype=True,
                                                       encoding=encoding)
+
+    if chunksize is not None:
+        _validate_integer("chunksize", chunksize, 1)
```
this should raise if ``lines!=True``

louispotok replied:
Will do.

louispotok replied:
Done

jreback commented on pandas/tests/io/json/test_pandas.py:
```diff
@@ -1032,6 +1032,32 @@ def test_to_jsonl(self):
         assert_frame_equal(pd.read_json(result, lines=True), df)
 
+    def test_read_jsonchunks(self):
+        df = pd.DataFrame({'A': [1, 2, 3], 'B': [4, 5, 6]})
```
add issue number as a comment

louispotok replied:
Will do.

louispotok replied:
Done

jreback commented on pandas/tests/io/json/test_pandas.py:
```diff
@@ -1032,6 +1032,32 @@ def test_to_jsonl(self):
+            pd.read_json(strio, lines=True, chunksize=2.2)
+
+        with tm.assert_raises_regex(ValueError, msg):
+            pd.read_json(strio, lines=True, chunksize='foo')
```
test with chunksize and lines=False

louispotok replied:
Will do.

louispotok replied:
Done

louispotok commented on doc/source/whatsnew/v0.21.0.txt:
```diff
@@ -197,6 +197,7 @@ Other API Changes
   raise an ``AttributeError`` rather than a ``ClosedFileError`` (:issue:`16301`)
 - :func:`read_csv` now treats ``'null'`` strings as missing values by default (:issue:`16471`)
 - :func:`read_csv` now treats ``'n/a'`` strings as missing values by default (:issue:`16078`)
+- :func:`read_json` now accepts a ``chunksize`` parameter that can reduce memory usage when ``lines=True``. (:issue:`17048`)
```
Will do.

louispotok commented:
It looks like we still need a decision on whether to return a `BaseIterator` or just yield chunks. I don't have a strong opinion on that and will wait for a decision there.

I'll point out one thing though. I think there are really two issues here worth separating:
1) The existing implementation doesn't permit loading JSONs even when the full dataframe will fit in memory (this was the original bug report).
2) Should we provide a way for users to handle JSON dataframes that *don't* fit in memory by returning an iterator / generator?

I was originally attempting to fix (1) but there may be another way to fix it instead. The culprit seems to be [line 349](https://github.com/pandas-dev/pandas/blob/master/pandas/io/json/json.py#L349) which on my laptop kills 10G of available memory with a 1.7GB json object. My solution to (1) was to read in the file in chunks, but maybe there are other possible fixes.  Then, I think a separate pull request could solve issue (2). 

What do you think?

jreback commented:
> It looks like we still need a decision on whether to return a BaseIterator or just yield chunks. I don't have a strong opinion on that and will wait for a decision there.

make a class that inherits BaseIterator, call it JsonLineReader. return this to the user. this solves both issues generally.

louispotok committed:
Return JsonLineReader from read_json

When chunksize is passed and lines=True, read_json
now returns a JsonLineReader, which inherits from
BaseIterator.
Also, internally, wrap up read_json kwargs into a dictionary
and pass them down opaquely to the SeriesParser or FrameParser.

louispotok committed:
Raise ValueError if chunksize is not None, but not lines

louispotok committed:
Add issue number to test docstring and add test.

Test that read_json raises exception if chunksize is passed
and lines != True.

louispotok committed:
bugfix: raise StopIteration, dont return it

pep8speaks commented:
Hello @louispotok! Thanks for updating the PR.

 - In the file [`pandas/io/json/json.py`](https://github.com/pandas-dev/pandas/blob/07b395f5bea64d2dbe80b39e2264fa37561e9fc2/pandas/io/json/json.py), following are the PEP8 issues :

> [Line 422:13](https://github.com/pandas-dev/pandas/blob/07b395f5bea64d2dbe80b39e2264fa37561e9fc2/pandas/io/json/json.py#L422): [E722](https://duckduckgo.com/?q=pep8%20E722) do not use bare except'

 - In the file [`pandas/tests/io/json/test_pandas.py`](https://github.com/pandas-dev/pandas/blob/07b395f5bea64d2dbe80b39e2264fa37561e9fc2/pandas/tests/io/json/test_pandas.py), following are the PEP8 issues :

> [Line 1063:9](https://github.com/pandas-dev/pandas/blob/07b395f5bea64d2dbe80b39e2264fa37561e9fc2/pandas/tests/io/json/test_pandas.py#L1063): [E303](https://duckduckgo.com/?q=pep8%20E303) too many blank lines (2)

louispotok committed:
PEP8 cleanup

louispotok commented:
*** Note: not ready for re-review -- found some bugs that were not caught by tests. Fixing behavior and tests.

louispotok committed:
Bugfixes for chunksize

* use _validate_integer to cast chunksize to integer, instead
  of just checking
* pass filepath_or_buffer to JsonLineReader where appropriate
* JsonLineReader uses integer index
* fix read_json chunksize tests to actually use StringIO and test Series

louispotok committed:
add chunksize test for reading from file

louispotok commented:
Okay, this is ready for re-review. 

@jreback I believe I've addressed all your comments.
@gfyoung In my latest (coarse) performance testing, with large enough chunks this is not noticeably slower than reading it all at once. Sorry for the false alarm.

One other point I want to confirm. I'm assuming (based on my reading of the code and documentation) that currently when `lines=True`, we always have `orient="records"` and the resulting object should always have an integer index `[0, 1, ... , len(obj) - 1 ]`. Is that a true assumption?

louispotok committed:
pep8 cleanup

louispotok commented:
I'm not sure why the AppVeyor build broke, it looks like an unrelated issue.

gfyoung commented:
@louispotok : Maybe not.  `ResourceWarning` could be indicative of a stream not having been properly closed, which is related to what you are doing.

louispotok committed:
Run chunksize checks before file is opened

Also, fix badly named vars in test.

louispotok commented:
Ah, I think I caught it. Let's see. To document fully, in my understanding, there are 4 cases to consider. Here's the current behavior:
1. filepath passed in without chunksize: file is opened, then read in, then closed (not changed from previous)
2. filepath passed in with chunksize: file is opened, then JsonLineReader is returned. When JsonLineReader gets to the end of the file, it attempts to close the file.
3. Buffer passed in without chunksize: json is read out of the buffer using filepath_or_buffer.read(), and then the object is returned, but the buffer stays open. (unchanged from previous)
4. Buffer passed in with chunksize: JsonLineReader is returned and assumes that buffer will stay open until it has finished iterating through. When it is done, it closes the buffer.

Do those decisions make sense? I could see the argument for changing (4) so that it leaves the buffer open and the caller has to close it.

I think (and we'll see with this build) the problem was in situation (2), I was checking that chunksize was valid *after* the file was opened, so the test suite correctly caught the error (`assert_raises_regex`) but was then exiting while leaving the file open. If that's true, we'll have to find a way to write a test for this.

louispotok commented:
Hm, still failing and I'm not sure how to investigate further. I do notice that the method that errors (`assert_produces_warning`) says it is not thread safe, not sure if that is relevant. Would appreciate some help here figuring out how to fix this.

louispotok commented:
@gfyoung 
I'm unable to reproduce this locally. If I were able to, I could try to trace it better (as in [here](https://docs.python.org/3/whatsnew/3.6.html?highlight=resourcewarning#warnings)) but I'm stuck. 

I've tried a number of things to get to the bottom of this but no luck:
1) Run the full test suite locally -- can't reproduce
2) Run the full test suite locally multithreaded as described [here](https://pandas.pydata.org/pandas-docs/stable/contributing.html#running-the-test-suite) -- can't reproduce
3) Created a test script to leave a bunch of files open (with `warnings.default_action = "error"`). This causes an IOError with ~1000 files open, but no ResourceWarning.
4) Created a test script to leave a bunch of buffers open (`StringIO` specifically, with `warnings.default_action = "error"), got into millions without a ResourceWarning.

I'd love any ideas or thoughts about how to proceed. Thanks!

gfyoung commented:
@louispotok : Sorry about that!  I think I missed your message in the sea of messages that I get daily 😄 .  I would suggest that you try rebasing onto `master` and see whether or not tests still fail.

I must say that it is kind of weird that this test was already failing from your first commit.  If AppVeyor still breaks, take that failing test and walk through the code using your changes from the your first commit *only* to see what's going on.

louispotok commented:
Great! @gfyoung I rebased to master and AppVeyor passes now. But the Travis CI build terminates because no output received in 10minutes, which again does not seem like a problem due to this change. WDYT?

gfyoung commented:
The OSX builds are generally a little more annoying on Travis.  I'll restart, as it terminates well-before it even builds `pandas`.

jreback commented:
yeah I here other folks (arrow) are having a similar Travis with macosx today.... 

cc @wesm

jreback commented on pandas/io/json/json.py:
```diff
@@ -320,8 +332,21 @@ def read_json(path_or_buf=None, orient=None, typ='frame', dtype=True,
 
+    if chunksize is not None:
+        chunksize = _validate_integer("chunksize", chunksize, 1)
+        if not lines:
```
check ``lines`` first

louispotok replied:
Done (I think I understood you correctly)

jreback replied:
this is ok i think, though should go inside jsonReader

jreback commented on pandas/tests/io/json/test_pandas.py:
```diff
@@ -1032,6 +1032,67 @@ def test_to_jsonl(self):
+
+        with tm.assert_raises_regex(ValueError, msg):
+            test_with_chunksize(-2.2)
+
```
rather than making a very long test, pls break into several tests, e.g. errors can be a test

louispotok replied:
Will do, thanks for the suggestion.

jreback commented on pandas/tests/io/json/test_pandas.py:
```diff
@@ -1032,6 +1032,67 @@ def test_to_jsonl(self):
+
+        def get_strio():
+            return StringIO(df.to_json(lines=True, orient="records"))
+
```
this is also hard to follow, pls use some comments to say what you are testing

louispotok replied:
Done

jreback commented on pandas/tests/io/json/test_pandas.py:
```diff
@@ -1032,6 +1032,67 @@ def test_to_jsonl(self):
+            return StringIO(df.to_json(lines=True, orient="records"))
+
+        def test_with_chunksize(c):
+            iterator = pd.read_json(get_strio(), lines=True, chunksize=c)
```
I don't really like the sub-functions.

Instead use a fixture for ``get_strio`` (and make it a more meaningful name.

louispotok replied:
Happy to rename it. 

I don't think I can use a fixture, though, because the `StringIO` that is returned can only be read once and every comparison needs to read it twice (chunked and unchunked). I'll move it into a separate function for now, but let me know what you think best.

louispotok commented on pandas/io/json/json.py:
```diff
@@ -334,12 +359,18 @@ def read_json(path_or_buf=None, orient=None, typ='frame', dtype=True,
-            json = fh.read()
-            fh.close()
+            if lines and chunksize:
+                return JsonLineReader(fh, chunksize, **kwargs)
```
If I do that:

1. I would need to repeat the [check for a `read` attribute](https://github.com/pandas-dev/pandas/pull/17168/files#diff-6c6379c395195326fa54aa784a1eb89fR369) on initialization.
2. I would need to pass down the `encoding`. 

I'd rather avoid repetition and preserve the separation of concerns -- what do you see as the benefit?

jreback replied:
encapsulation is most important here and you have a lot of code complexity currently

1) not sure that is necessary and if so it's cheap 
2) again that is fine

we have this pattern in all other readers 
so pls make the change

louispotok replied:
Thanks for the explanation. Done, let me know what you think.

louispotok committed:
move strio df in test to fixture

louispotok committed:
Improve read_json chunking tests

* split into multiple tests
* move helper functions out
* add comments

louispotok committed:
bugfix in read_json tests, remove fixture

louispotok commented:
@jreback Thanks for the review. I incorporated some of your comments. Two places I diverged:
1. Using a function instead of a fixture for the test data, since I need a StringIO and if I use a fixture I can only read it once per test.
2. I didn't make the suggested change around opening the file within `JsonLineReader`, explanation [here](https://github.com/pandas-dev/pandas/pull/17168#discussion_r137959567).

louispotok committed:
JsonLineReader opens and closes filepaths

louispotok committed:
pep8 cleanup

louispotok commented:
@jreback This is ready for another look whenever you have a chance.

jreback commented on pandas/io/json/json.py:
```diff
@@ -349,22 +382,71 @@ def read_json(path_or_buf=None, orient=None, typ='frame', dtype=True,
         lines = list(StringIO(json.strip()))
         json = '[' + ','.join(lines) + ']'
 
+    return _get_obj(json, **kwargs)
```
this doesn't need to be a separate function

jreback commented on pandas/io/json/json.py:
```diff
@@ -349,22 +382,71 @@ def read_json(path_or_buf=None, orient=None, typ='frame', dtype=True,
+
+            return obj
+
+        else:
```
The close can be its own method

jreback commented on pandas/io/json/json.py:
```diff
@@ -349,22 +382,71 @@ def read_json(path_or_buf=None, orient=None, typ='frame', dtype=True,
+        self.nrows_seen = 0
+
+    def __next__(self):
+        lines = list(islice(self.iterator, self.chunksize))
```
is lines possible to be empty?

louispotok replied:
Yes, it's empty after all the lines have been read, which triggers the StopIteration.

jreback commented on pandas/tests/io/json/test_pandas.py:
```diff
@@ -35,6 +35,15 @@
 _mixed_frame = _frame.copy()
 
 
+def strio_lines_json_df():
```
make this a fixture

jreback commented on pandas/tests/io/json/test_pandas.py:
```diff
@@ -1032,6 +1041,72 @@ def test_to_jsonl(self):
         assert result == expected
         assert_frame_equal(pd.read_json(result, lines=True), df)
 
+    def test_readjson_chunks(self):
```
you input the fixture as an arg here

jreback commented on pandas/tests/io/json/test_pandas.py:
```diff
@@ -35,6 +35,15 @@
+    return StringIO(df.to_json(lines=True, orient="records"))
+
+
+def json_lines_to_df_chunked(jlines, chunksize):
```
I wouldn't make this a function at all, just in-line it, its very confusing to the flow otherwise

jreback commented on pandas/tests/io/json/test_pandas.py:
```diff
@@ -1032,6 +1041,72 @@ def test_to_jsonl(self):
+        read_json(chunks=False)"""
+        # GH17048: memory usage when lines=True
+
+        unchunked = pd.read_json(strio_lines_json_df(), lines=True)
```
use parametrize with the different args here

jreback commented on pandas/tests/io/json/test_pandas.py:
```diff
@@ -1032,6 +1041,72 @@ def test_to_jsonl(self):
+            chunked = pd.concat(pd.read_json(path, lines=True, chunksize=1))
+            unchunked = pd.read_json(path, lines=True)
+            assert_frame_equal(unchunked, chunked)
+
```
use parametrize

louispotok replied:
done

jreback commented on pandas/io/json/json.py:
```diff
@@ -332,14 +357,22 @@ def read_json(path_or_buf=None, orient=None, typ='frame', dtype=True,
 
         if exists:
-            fh, handles = _get_handle(filepath_or_buffer, 'r',
-                                      encoding=encoding)
```
So I would change this around a bit. Make this the JsonReader. and pass it both lines and chunksize (as well as all the other params).

So you *always* create it.

Then if you have lines & chunksize you can return an iterator, otherwise you can simply return it like this

```
jr = JsonReader(.....)


if lines & chunksize:
     return jr

return jr.read()
```

Then you can make ``_get_obj`` a method on JsonReader. Much cleaner this way.

This is much more like how ``pandas.io.pytables.TableIterator`` (and read_csv iterator) are setup. You *always* create a reader, and either return it or just call .read() on it.

louispotok committed:
update whatsnew

louispotok committed:
update docs on read_json chunksize

louispotok committed:
Always use JsonReader in read_json

Either read it all before return, or return the JsonReader if chunksize
is passed.

louispotok committed:
make lines_json_df a fixture

louispotok committed:
remove unneeded concats in tests

louispotok committed:
parametrize some tests

louispotok commented:
@jreback Thanks for the helpful review. New changes:
* updated the docs
* use JsonReader in both cases (and just `return JsonReader.read()` when appropriate)
* Created the test fixture and removed the other testing helper function as per request

Two things you asked that I wasn't sure about:
* Moving close to its own method -- happy to do it if you like, but is there another advantage to it?
* I parameterized the tests in two places by just using a for loop. Given what I'm doing and how the fixture is constructed I'm not sure how using pytests fixture parametrization would actually work here, so if you had something else in mind I'd appreciate a more explicit instruction.

jreback commented:
> Moving close to its own method -- happy to do it if you like, but is there another advantage to it?

just hygiene

jreback commented on pandas/io/json/json.py:
```diff
@@ -320,49 +334,140 @@ def read_json(path_or_buf=None, orient=None, typ='frame', dtype=True,
+    if chunksize:
+        return json_reader
+
     else:
```
dont need the else

louispotok replied:
removed

jreback commented on pandas/io/json/json.py:
```diff
@@ -320,49 +334,140 @@ def read_json(path_or_buf=None, orient=None, typ='frame', dtype=True,
+            return self._get_obj(self.data)
+        elif self.lines and self.chunksize:
+            return concat(self)
+        else:
```
don't need the else

louispotok replied:
Moved this around a little bit to only have one return statement, so we can call `self.close()` in only one place. The else is now needed.

jreback commented on pandas/io/json/json.py:
```diff
@@ -320,49 +334,140 @@ def read_json(path_or_buf=None, orient=None, typ='frame', dtype=True,
+            self.raw_json = True
+            self.data = filepath_or_buffer
+
+        if self.raw_json and lines:
```
why are you doing combine lines here?

louispotok replied:
Moved it out of `_init_` and into `read`, adding another conditional.

jreback commented on pandas/io/json/json.py:
```diff
@@ -320,49 +334,140 @@ def read_json(path_or_buf=None, orient=None, typ='frame', dtype=True,
+            else:
+                return self._get_obj(self.data.read())
+
+    def _get_obj(self, json):
```
add a doc-string, maybe rename to _get_object_parser

louispotok replied:
done

jreback commented on pandas/io/json/json.py:
```diff
@@ -320,49 +334,140 @@ def read_json(path_or_buf=None, orient=None, typ='frame', dtype=True,
+
+            return obj
+
+        else:
```
don't need the else

louispotok replied:
removed it

jreback commented on pandas/tests/io/json/test_pandas.py:
```diff
@@ -1032,6 +1038,65 @@ def test_to_jsonl(self):
         assert result == expected
         assert_frame_equal(pd.read_json(result, lines=True), df)
 
+    def test_readjson_chunks(self, lines_json_df):
```
can you add a test which writes a file in line-format and reads it back. check that the file handle is closed properly.

louispotok replied:
done

jreback reviewed (changes requested):
couple of comments. can you add an asv benchmark that uses (lines=True), or if we have one, pls run and report results. (do with and w/o a chunksize)

jreback commented:
@chris-b1 can you have a look.

gfyoung commented on doc/source/io.rst:
```diff
@@ -2059,6 +2064,12 @@ using Hadoop or Spark.
   df.to_json(orient='records', lines=True)
 
+  # chunksize has no effect when reading a string.
+  import io
```
Also, it's the wrong library.  It should be `pandas.compat.StringIO`.

louispotok replied:
Ah yes, I see it's imported at the top. Thanks, fixed.

gfyoung commented on pandas/io/json/json.py:
```diff
@@ -320,49 +334,140 @@ def read_json(path_or_buf=None, orient=None, typ='frame', dtype=True,
+        self, filepath_or_buffer, orient, typ, dtype, convert_axes,
+        convert_dates, keep_default_dates, numpy, precise_float, date_unit,
+        encoding, lines, chunksize, raw_json=False
+    ):
```
I think we can move this parenthesis up to the previous line.  This looks more Java than Python :smile:

louispotok replied:
done

gfyoung commented on pandas/io/json/json.py:
```diff
@@ -320,49 +334,140 @@ def read_json(path_or_buf=None, orient=None, typ='frame', dtype=True,
+                exists = os.path.exists(filepath_or_buffer)
+
+            # if the filepath is too long will raise here
+            # 5874
```
Put this as one line:
~~~
gh-5874: if the filepath is too long, raise here
~~~

louispotok replied:
done

gfyoung commented on pandas/io/json/json.py:
```diff
@@ -320,49 +334,140 @@ def read_json(path_or_buf=None, orient=None, typ='frame', dtype=True,
+            lines_json = '[' + ','.join(lines) + ']'
+            obj = self._get_obj(lines_json)
+
+            # Make sure that the returned objects have the right index
```
Nit: period at the end.

louispotok replied:
done

louispotok committed:
add __close__ method to JsonReader and use it

louispotok committed:
remove import io in docs

louispotok committed:
move read_json in whatsnew to Other Enhancements

louispotok committed:
move chunksize and lines validation into JsonReader

louispotok committed:
remove extraneous else

louispotok committed:
remove unneccessary cast to list

louispotok committed:
move combine_lines call into read

louispotok committed:
remove another extraneous else

louispotok committed:
always close JsonReader

louispotok committed:
add test that read_json closes file correctly

louispotok committed:
minor formatting fixups

louispotok commented on doc/source/whatsnew/v0.21.0.txt:
```diff
@@ -321,6 +321,7 @@ Other API Changes
   raise an ``AttributeError`` rather than a ``ClosedFileError`` (:issue:`16301`)
 - :func:`read_csv` now treats ``'null'`` strings as missing values by default (:issue:`16471`)
 - :func:`read_csv` now treats ``'n/a'`` strings as missing values by default (:issue:`16078`)
+- :func:`read_json` now accepts a ``chunksize`` parameter that can be used when ``lines=True``. If ``chunksize`` is passed, read_json now returns an iterator which reads in ``chunksize`` lines with each iteration. (:issue:`17048`)
```
done

louispotok committed:
remove extraneous else

louispotok committed:
add benchmarks for read_json

louispotok committed:
update benchmarks

louispotok committed:
move json_lines tests to io_bench

louispotok committed:
add peakmem for jsonlines

louispotok committed:
smaller benchmark

louispotok commented:
Made all the requested fixes and added a benchmark for `lines=True` with and without a chunksize.
@jreback on my machine this doesn't change the time performance of `lines=True` without chunksize, and actually it looks slightly faster with chunksize passed.

chris-b1 commented:
This seems OK to me - only surprising behavior (you note it in the docs) is that this doesn't work with StringIO - doesn't seem like there is any technical reason it has to be that way?  E.g. read_csv handles it and chunksize.

```python
In [17]: i = pd.read_csv(StringIO("""a,b
    ...: 1,2
    ...: 3,4
    ...: 5,6
    ...: 7,8"""), chunksize=2)

In [18]: for d in i:
    ...:     print(d)
   a  b
0  1  2
1  3  4
   a  b
2  5  6
3  7  8
```

jreback commented on asv_bench/benchmarks/io_bench.py:
```diff
@@ -192,3 +192,25 @@ def time_read_nrows(self, compression, engine):
+        self.N = 100000
+        self.C = 5
+        self.df = DataFrame(dict([('float{0}'.format(i), randn(self.N)) for i in range(self.C)]))
+        self.df.to_json("__test__.json",orient="records",lines=True)
```
add a ``teardown`` method to remove the created file here (you can name it like)

```
class read_json_lines(...):
      file = '__test__.json'

      def teardown(self):
             try:
                  os.remove(self.file)
             except:
                  pass
```

jreback replied:
I think we need to scour the asv's for not-using teardowns as well (almost all prob in io benchmarks). @louispotok if you could add would be great (or could do in another issue)

louispotok replied:
Created a new issue [here](https://github.com/pandas-dev/pandas/issues/17616), I'm happy to do it.

louispotok replied:
And, created the teardown method.

jreback commented on pandas/io/json/json.py:
```diff
@@ -322,47 +336,143 @@ def read_json(path_or_buf=None, orient=None, typ='frame', dtype=True,
+                self.data, _ = _get_handle(filepath_or_buffer, 'r',
+                                           encoding=encoding)
+            else:
+                self.raw_json = True
```
why do we care about this ``.raw_json`` flag? esp in the constructor?

jreback commented on pandas/io/json/json.py:
```diff
@@ -322,47 +336,143 @@ def read_json(path_or_buf=None, orient=None, typ='frame', dtype=True,
+    def __next__(self):
+        lines = list(islice(self.data, self.chunksize))
+        if lines:
+            lines_json = '[' + ','.join(lines) + ']'
```
why are you not using the combine_lines here?

jreback commented on pandas/io/json/json.py:
```diff
@@ -322,47 +336,143 @@ def read_json(path_or_buf=None, orient=None, typ='frame', dtype=True,
+                obj = self._get_object_parser(self.combine_lines(self.data))
+            else:
+                obj = self._get_object_parser(self.data)
+        elif self.lines and self.chunksize:
```
you need to check lines & chunksize first I think, then determine if its raw_json here

jreback commented on pandas/io/json/json.py:
```diff
@@ -322,47 +336,143 @@ def read_json(path_or_buf=None, orient=None, typ='frame', dtype=True,
+
+    def read(self):
+        """Read the whole JSON input into a pandas object"""
+        if self.raw_json:
```
this should be your last condition (IOW the else)

jreback reviewed (changes requested):
I think need to work on the .read(), maybe something like

```
if self.lines and self.chunksize:
    ....
elif self.lines:
    ...
else:
    .....read the whole thing
```

louispotok commented:
@chris-b1 Ah, good catch. Those docs are super unclear, what I meant was "it works with a StringIO but not with a string, so we'll wrap `jsonl` in StringIO to show what it does."
I'll fix it, but since you mention it: do you think we should support strings as well? From the user's perspective there's no advantage (since the whole string will be in memory anyway) but it simplifies the API a little bit. WDYT?

louispotok committed:
refactor JsonReader

louispotok committed:
add test for reading with multiple empty lines

louispotok committed:
add support for JSON docs with multiple consecutive newlines

louispotok committed:
remove raw_json init param

louispotok committed:
DRY for combining lines

louispotok committed:
use floor division in asv bench

louispotok committed:
add teardown to asv bench

louispotok commented on asv_bench/benchmarks/io_bench.py:
```diff
@@ -192,3 +192,25 @@ def time_read_nrows(self, compression, engine):
+    def time_read_json_lines(self):
+        pd.read_json("__test__.json", lines=True)
+
+    def time_read_json_lines_chunk(self):
```
Yes, we do: `test_readjson_invalid_chunksize`. Changed to `//`.

louispotok committed:
add docs

louispotok commented:
@jreback Thanks, agree that the code was a little convoluted. I've made some changes that hopefully make everything clearer.

louispotok committed:
pep fixup

jreback commented on doc/source/io.rst:
```diff
@@ -2059,6 +2064,11 @@ using Hadoop or Spark.
   df
   df.to_json(orient='records', lines=True)
 
+  # chunksize has no effect when reading a string.
```
this is an odd comment here
it does work and does iterate; you are referring to memory savings which is not relevant here

louispotok replied:
You're right. @chris-b1 commented on the same line, I was waiting for a response from him before changing it. Do you agree that we should support chunksize when a string is passed? Either way I will change the documentation.

chris-b1 replied:
I didn't know we would fall back to parsing the string if not a file (arguably shouldn't, but whole other issue - xref #5924)

So I'm fine not supporting chunksize with a string, but would like this comment to be clearer, or removed.

jreback commented on pandas/io/json/json.py:
```diff
@@ -263,6 +266,17 @@ def read_json(path_or_buf=None, orient=None, typ='frame', dtype=True,
+        for more information on ``chunksize``.
+        This can only be passed if `lines=True`.
+        If this is None, the file will be read into memory all at once.
+        If the JSON input is a string, this argument has no effect.
```
this last line is incorrect

louispotok replied:
True, will fix.

jreback commented on pandas/io/json/json.py:
```diff
@@ -322,47 +336,178 @@ def read_json(path_or_buf=None, orient=None, typ='frame', dtype=True,
+    Reads a JSON document to a pandas object.
+
+    If initialized with ``lines=True`` and ``chunksize``, can be iterated over
+    ``chunksize`` lines at a time. Otherwise, calling ``read`` reads in the
```
jsonreader provides an iterator interface for reading

louispotok replied:
changed to "JsonReader provides an interface for reading in a JSON file.", since it is only an iterator interface if `lines=True` and `chunksize!=None`.

jreback commented on pandas/io/json/json.py:
```diff
@@ -322,47 +336,178 @@ def read_json(path_or_buf=None, orient=None, typ='frame', dtype=True,
+            except (TypeError, ValueError):
+                exists = False
+
+            if exists:
```
this can be written as an else to the try/except
you don't need the exists flag

louispotok replied:
Fixed.

jreback commented on pandas/io/json/json.py:
```diff
@@ -322,47 +336,178 @@ def read_json(path_or_buf=None, orient=None, typ='frame', dtype=True,
+                data, _ = _get_handle(filepath_or_buffer, 'r',
+                                      encoding=self.encoding)
+
+        if not data:
```
you can assign this to data initially
it will get overwritten if the try/except succeeds

louispotok replied:
Fixed.

jreback commented on pandas/io/json/json.py:
```diff
@@ -322,47 +336,178 @@ def read_json(path_or_buf=None, orient=None, typ='frame', dtype=True,
+        Otherwise, we read it into memory for the `read` method.
+        """
+        if hasattr(data, 'read'):
+            if self.chunksize:
```
this can be simpler

louispotok replied:
I simplified it into two "if" statements, not sure if it's more or less readable this way. LMK what you think:

```
def _preprocess_data(self, data): 
    if hasattr(data, 'read') and not self.chunksize:                         
        data = data.read()                                                      
    if not hasattr(data, 'read') and self.chunksize:                            
        data = StringIO(data)                                                   
                                                                                    
    return data
```

jreback commented on pandas/tests/io/json/test_pandas.py:
```diff
@@ -1032,6 +1038,102 @@ def test_to_jsonl(self):
+        read_json(chunks=False)"""
+        # GH17048: memory usage when lines=True
+
+        for cs in [1, 1.0]:
```
a float is not a valid chunksize or do we allow this if it's equal to an int?
we shouldn't do that

louispotok replied:
We use _validate_integer to be consistent with the API for [chunksize in read_csv](https://github.com/pandas-dev/pandas/blob/master/pandas/io/parsers.py#L407).

jreback commented on pandas/tests/io/json/test_pandas.py:
```diff
@@ -1032,6 +1038,102 @@ def test_to_jsonl(self):
+
+        for cs in [1, 1.0]:
+
+            unchunked = pd.read_json(StringIO(lines_json_df), lines=True)
```
if yes to the above then parametrize this test with cs (but call it chunksize)

louispotok replied:
done

jreback commented on pandas/tests/io/json/test_pandas.py:
```diff
@@ -1032,6 +1038,102 @@ def test_to_jsonl(self):
         assert result == expected
         assert_frame_equal(pd.read_json(result, lines=True), df)
```
the lines=Trye test could go in a separate test class (you can move the existing ones)

louispotok replied:
Will do. While we're at it, what do you think of the name of the existing test class (`TestPandasContainer`)? I want them to be consistent, but I don't really understand the name of the existing one.

jreback replied:
sure i think that was copied from the actual json tests (so should be more reflective);

also ok with a separate file and no class (iow bare functions); this was just the existing structure ; we can/should use fixtures

jreback commented on pandas/tests/io/json/test_pandas.py:
```diff
@@ -1032,6 +1038,102 @@ def test_to_jsonl(self):
+
+    def test_readjson_each_chunk(self, lines_json_df):
+        """
+        Other tests check that the final result of read_json(chunksize=True) is
```
don't use quotes for test use # for comments

louispotok replied:
done

jreback commented on pandas/tests/io/json/test_pandas.py:
```diff
@@ -1032,6 +1038,102 @@ def test_to_jsonl(self):
+                if chunksize is not None:
+                    pd.concat(pd.read_json(f, lines=True, chunksize=chunksize))
+                else:
+                    pd.read_json(f, lines=True)
```
this is not correct
we should only close the file if we open it ourselves
if we have an open file it should stay open

louispotok replied:
Ah, that makes sense and is in line with current behavior. So in [this comment](https://github.com/pandas-dev/pandas/pull/17168/files#r139048920), did you mean something like the following?

```
           df = pd.DataFrame({'A': [1, 2, 3], 'B': [4, 5, 6]})                      
            df.to_json(path, lines=True, orient="records")                           
            reader = JsonReader(                                                     
                path, orient=None, typ="frame", dtype=True, convert_axes=True,   
                convert_dates=True, keep_default_dates=True, numpy=False,        
                precise_float=False, date_unit=None, encoding=None,              
                lines=True, chunksize=chunksize)                                 
            reader.read()                                                        
            assert reader.open_stream.closed
```

louispotok committed:
update documentation

louispotok committed:
simplify JsonReader._preprocess_data

louispotok committed:
simplify _get_data_from_filepath

louispotok committed:
Update read_json tests

Split out tests with lines=True into separate test class
Parametrize tests
Replace """ comments with #.

louispotok committed:
JsonReader should only close if it opened

louispotok committed:
split out json readlines to sep test class

louispotok committed:
add encoding to test_readlines

louispotok committed:
pep8 cleanup

jreback commented:
this looks good. can you show the asv results vs master? (of course the chunksize won't work on master), but IIRC your example will allow us to compare them vs the non-chunksize ones.

jreback commented on doc/source/io.rst:
```diff
@@ -2059,6 +2064,10 @@ using Hadoop or Spark.
   df
   df.to_json(orient='records', lines=True)
 
+  reader = pd.read_json(StringIO(jsonl), lines=True, chunksize=1)
```
maybe a comment here that reader is an iterator

louispotok replied:
done

jreback commented on pandas/io/json/json.py:
```diff
@@ -323,47 +336,167 @@ def read_json(path_or_buf=None, orient=None, typ='frame', dtype=True,
+        data = self._get_data_from_filepath(filepath_or_buffer)
+        self.data = self._preprocess_data(data)
+
+    def _preprocess_data(self, data):
```
you could move _preprocess_data back to ``__init__``

louispotok replied:
I prefer it split out, I think it keeps things easier to read. But I can move it back if you like.

jreback commented on pandas/io/json/json.py:
```diff
@@ -323,47 +336,167 @@ def read_json(path_or_buf=None, orient=None, typ='frame', dtype=True,
+
+        return data
+
+    def combine_lines(self, lines):
```
make this private (``_combine_lines``)

louispotok replied:
done

louispotok committed:
minor fixups

louispotok commented:
@jreback Great! Glad to be ready to ship this. Do you want me to squash into one/a few commits first?

Relevant asv benchmarks:
```
(pandas_dev) louispotok@oye:~/pandas-louispotok/asv_bench$ asv continuous -f 1.1 upstream/master HEAD -b io_bench.read_json_lines
· Creating environments
· Discovering benchmarks
·· Uninstalling from conda-py3.6-Cython-matplotlib-numexpr-numpy-openpyxl-pytables-pytest-scipy-sqlalchemy-xlrd-xlsxwriter-xlwt
·· Installing into conda-py3.6-Cython-matplotlib-numexpr-numpy-openpyxl-pytables-pytest-scipy-sqlalchemy-xlrd-xlsxwriter-xlwt..
· Running 8 total benchmarks (2 commits * 1 environments * 4 benchmarks)
[  0.00%] · For pandas commit hash c84e5154:
[  0.00%] ·· Building for conda-py3.6-Cython-matplotlib-numexpr-numpy-openpyxl-pytables-pytest-scipy-sqlalchemy-xlrd-xlsxwriter-xlwt..................................................
[  0.00%] ·· Benchmarking conda-py3.6-Cython-matplotlib-numexpr-numpy-openpyxl-pytables-pytest-scipy-sqlalchemy-xlrd-xlsxwriter-xlwt
[ 12.50%] ··· Running io_bench.read_json_lines.peakmem_read_json_lines                           185M
[ 25.00%] ··· Running io_bench.read_json_lines.peakmem_read_json_lines_chunk                     131M
[ 37.50%] ··· Running io_bench.read_json_lines.time_read_json_lines                           272±1ms
[ 50.00%] ··· Running io_bench.read_json_lines.time_read_json_lines_chunk                     285±2ms
[ 50.00%] · For pandas commit hash d3be81ad:
[ 50.00%] ·· Building for conda-py3.6-Cython-matplotlib-numexpr-numpy-openpyxl-pytables-pytest-scipy-sqlalchemy-xlrd-xlsxwriter-xlwt.................................................
[ 50.00%] ·· Benchmarking conda-py3.6-Cython-matplotlib-numexpr-numpy-openpyxl-pytables-pytest-scipy-sqlalchemy-xlrd-xlsxwriter-xlwt
[ 62.50%] ··· Running io_bench.read_json_lines.peakmem_read_json_lines                           202M
[ 75.00%] ··· Running io_bench.read_json_lines.peakmem_read_json_lines_chunk                   failed
[ 87.50%] ··· Running io_bench.read_json_lines.time_read_json_lines                         273±0.6ms
[100.00%] ··· Running io_bench.read_json_lines.time_read_json_lines_chunk                      failed
SOME BENCHMARKS HAVE CHANGED SIGNIFICANTLY.
```

jreback commented:
can you rebase; you don't need to squash

jreback commented:
thanks @louispotok nice PR! very responsive! keep em coming!

louispotok commented:
This was fun, thanks for all your help!
